gnovm/stdlibs/std/coins.gno

```diff
@@ -13,28 +14,125 @@ type Coin struct {
 	Amount int64  `json:"amount"`
 }
 
+// NewCoin returns a new coin with a denomination and amount
+func NewCoin(denom string, amount int64) Coin {
+	return Coin{
+		Denom:  denom,
+		Amount: amount,
+	}
+}
+
+// String provides a human-readable representation of a coin
 func (c Coin) String() string {
 	return strconv.Itoa(int(c.Amount)) + c.Denom
 }
 
+// IsGTE returns true if they are the same type and the receiver is
+// an equal or greater value
 func (c Coin) IsGTE(other Coin) bool {
-	if c.Denom != other.Denom {
-		panic("invalid coin denominations: " + c.Denom)
-	}
+	mustMatchDenominations(c.Denom, other.Denom)
+
 	return c.Amount >= other.Amount
 }
 
+// IsLT returns true if they are the same type and the receiver is
+// a smaller value
+func (c Coin) IsLT(other Coin) bool {
+	mustMatchDenominations(c.Denom, other.Denom)
+
+	return c.Amount < other.Amount
+}
+
+// IsEqual returns true if the two sets of Coins have the same value
+func (c Coin) IsEqual(other Coin) bool {
+	mustMatchDenominations(c.Denom, other.Denom)
+
+	return c.Amount == other.Amount
+}
+
+// Adds amounts of two coins with same denom.
+// If the coins differ in denom then it panics.
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Add(coinB Coin) Coin {
+	res := c.AddUnsafe(coinB)
+	return res
+}
+
+func (c Coin) AddUnsafe(coinB Coin) Coin {
+	mustMatchDenominations(c.Denom, coinB.Denom)
+
+	sum, ok := overflow.Add64(c.Amount, coinB.Amount)
+	if !ok {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+
+	c.Amount = sum
+	return c
+}
+
+// Subtracts amounts of two coins with same denom.
+// If the coins differ in denom then it panics.
+// An overflow or underflow panics.
+// An invalid result panics.
+func (c Coin) Sub(coinB Coin) Coin {
+	res := c.SubUnsafe(coinB)
+	return res
+}
+
+func (c Coin) SubUnsafe(coinB Coin) Coin {
+	mustMatchDenominations(c.Denom, coinB.Denom)
+
+	dff, ok := overflow.Sub64(c.Amount, coinB.Amount)
+	if !ok {
+		panic("coin add overflow/underflow: " + strconv.Itoa(int(c.Amount)) + " + " + strconv.Itoa(int(coinB.Amount)))
+	}
+	c.Amount = dff
+
+	return c
+}
+
+// IsPositive returns true if coin amount is positive.
+func (c Coin) IsPositive() bool {
+	return c.Amount > 0
+}
+
+// IsNegative returns true if the coin amount is negative and false otherwise.
+func (c Coin) IsNegative() bool {
```

Comment:
I have questions on the validity of IsNegative/IsZero, but I know this is just copying the code from the existing implementation, so it's OK for now